BepInEx.Core/Bootstrap/BaseChainloader.cs

```diff
@@ -331,117 +347,127 @@ public virtual void Execute()
             {
-                static bool IsHardDependency(BepInDependency dep) =>
-                    (dep.Flags & BepInDependency.DependencyFlags.HardDependency) != 0;
-
-                // If the dependency wasn't already processed, it's missing altogether
-                var dependencyExists =
-                    processedPlugins.TryGetValue(dependency.DependencyGUID, out var pluginVersion);
-                // Alternatively, if the dependency hasn't been loaded before, it's missing too
-                if (!dependencyExists)
+                continue;
+            }
+
+            try
+            {
+                Logger.Log(LogLevel.Info, $"Loading [{plugin}]");
+
+                Assembly ass = null;
+                if (provider && !loadedAssemblies.TryGetValue(plugin.Location, out ass))
                 {
-                    dependencyExists = Plugins.TryGetValue(dependency.DependencyGUID, out var pluginInfo);
-                    pluginVersion = pluginInfo?.Metadata.Version;
+                    ass = Assembly.LoadFrom(plugin.Location);
+                    loadedAssemblies[plugin.Location] = ass;
                 }
-
-                if (!dependencyExists || dependency.VersionRange != null &&
-                    !dependency.VersionRange.IsSatisfied(pluginVersion))
+                else if (!provider && !loadedAssemblies.TryGetValue(plugin.Loader.AssemblyIdentifier, out ass))
                 {
-                    // If the dependency is hard, collect it into a list to show
-                    if (IsHardDependency(dependency))
-                        missingDependencies.Add(dependency);
-                    continue;
+                    ass = Assembly.Load(plugin.Loader.GetAssemblyData());
+                    loadedAssemblies[plugin.Loader.AssemblyIdentifier] = ass;
                 }
 
-                // If the dependency is a hard and is invalid (e.g. has missing dependencies), report that to the user
-                if (invalidPlugins.Contains(dependency.DependencyGUID) && IsHardDependency(dependency))
+                existingPlugins[plugin.Metadata.GUID] = plugin;
+                TryRunModuleCtor(plugin, ass);
+
+                if (provider)
                 {
-                    dependsOnInvalidPlugin = true;
-                    break;
+                    var type = ass.GetType(plugin.TypeName);
+                    plugin.Instance = Activator.CreateInstance(type);
+                    AppDomain.CurrentDomain.AssemblyResolve += (_, args) => ((BasePluginProvider)plugin.Instance).ResolveAssembly(args.Name);
+                    var requestedPluginsStream = ((BasePluginProvider)plugin.Instance).GetPlugins();
+                    Loaders.AddRange(requestedPluginsStream);
                 }
+                else
+                {
+                    plugin.Instance = LoadPlugin(plugin, ass);
+                }
+                PluginLoaded?.Invoke(plugin);
```

Comment:
Should the event be called even for providers? Maybe it should include info that it is a provider or have a separate event?

Reply:
Added a separate event